File: dotnet/src/Microsoft.AutoGen/AgentChat/GroupChat/GroupChatManagerBase.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// GroupChatManagerBase.cs

using Microsoft.AutoGen.AgentChat.Abstractions;
using Microsoft.AutoGen.Contracts;

namespace Microsoft.AutoGen.AgentChat.GroupChat;

public abstract class GroupChatManagerBase : IGroupChatHandler
{
    private GroupChatOptions options;

    // TODO: We may be able to abstract this out at the Core level
    private MessagePublishServicer? PublishServicer { get; set; }

    // It would be so awesome if we could avoid passing GroupChatOptions to the constructor
    // and use something like Python's context manager mechanism to pick up the options from
    // the logical stack. But that's very difficult in C#, because the user code could do all
    // sorts of weird things like shunt the execution to a different thread. We cannot even
    // assume that we are in an async context (much less that we are in the same async context)
    public GroupChatManagerBase(GroupChatOptions options) : base()
    {
        this.options = options;

        this.MessageThread = new List<AgentMessage>();
    }

    protected string GroupChatTopicType => this.options.GroupChatTopicType;
    protected string OutputTopicType => this.options.OutputTopicType;

    protected Dictionary<string, GroupParticipant> Participants => this.options.Participants;

    protected ITerminationCondition? TerminationCondition => this.options.TerminationCondition;
    protected int? MaxTurns => this.options.MaxTurns;

    protected int CurrentTurn { get; set; }

    protected List<AgentMessage> MessageThread;

    void IGroupChatHandler.AttachMessagePublishServicer(MessagePublishServicer? servicer)
    {
        this.PublishServicer = servicer;
    }

    private ValueTask PublishMessageAsync(GroupChatEventBase message, string topicType, CancellationToken cancellation = default)
    {
        return this.PublishServicer?.Invoke(message, topicType, cancellation) ?? ValueTask.CompletedTask;
    }

    protected ValueTask PublishMessageAsync(ChatMessage message, string topicType, CancellationToken cancellation = default)
    {
        return this.PublishMessageAsync(new GroupChatMessage { Message = message }, topicType, cancellation);
    }

    protected virtual async ValueTask ValidateGroupState(List<ChatMessage>? messages)
    {
    }

    public abstract ValueTask<string> SelectSpeakerAsync(List<AgentMessage> thread);

    public async ValueTask HandleAsync(GroupChatStart item, MessageContext messageContext)
    {
        if (this.TerminationCondition != null && this.TerminationCondition.IsTerminated)
        {
            // skipReset is used here to match the Python code
            await this.TerminateAsync("The chat has already terminated", skipReset: true);

            StopMessage earlyStop = new StopMessage
            {
                Content = "The chat has already terminated",
                Source = GroupChatBase<GroupChatManagerBase>.GroupChatManagerTopicType
            };

            await this.PublishMessageAsync(new GroupChatTermination { Message = earlyStop }, this.OutputTopicType);

            return;
        }

        if (item.Messages != null)
        {
            this.MessageThread.AddRange(item.Messages);
        }

        await this.ValidateGroupState(item.Messages);

        if (item.Messages != null)
        {
            await this.PublishMessageAsync(item, this.OutputTopicType);
            await this.PublishMessageAsync(item, this.GroupChatTopicType);

            // item.Messages is IList<ChatMessage> but we need IList<AgentMessage>
            // Unfortunately, IList does not support type variance, so we have to do this rather ugly thing
            // TODO: Check if we really need to have AgentMessage on the interface of ITerminationCondition
            List<AgentMessage> converted = [.. item.Messages.Cast<AgentMessage>()];

            if (await this.TerminateIfNeededAsync(converted))
            {
                return;
            }
        }

        await this.ProcessNextSpeakerAsync();
    }

    public async ValueTask HandleAsync(GroupChatAgentResponse item, MessageContext messageContext)
    {
        List<AgentMessage> delta = new List<AgentMessage>();

        if (item.AgentResponse.InnerMessages != null)
        {
            this.MessageThread.AddRange(item.AgentResponse.InnerMessages);
            delta.AddRange(item.AgentResponse.InnerMessages);
        }

        this.MessageThread.Add(item.AgentResponse.Message);
        delta.Add(item.AgentResponse.Message);

        if (await this.TerminateIfNeededAsync(delta))
        {
            return;
        }

        this.CurrentTurn++;
        if (this.MaxTurns.HasValue && this.MaxTurns.Value <= this.CurrentTurn)
        {
            await this.TerminateAsync($"Maximum number of turns ({this.MaxTurns.Value}) reached.");
            return;
        }

        await this.ProcessNextSpeakerAsync();
    }

    private ValueTask TerminateAsync(string message, bool skipReset = false)
    {
        StopMessage stopMessage = new StopMessage
        {
            Content = message,
            Source = GroupChatBase<GroupChatManagerBase>.GroupChatManagerTopicType
        };

        return this.TerminateAsync(stopMessage, skipReset);
    }

    private async ValueTask TerminateAsync(StopMessage stopMessage, bool skipReset = false)
    {
        await this.PublishMessageAsync(new GroupChatTermination { Message = stopMessage }, this.OutputTopicType);

        if (!skipReset)
        {
            this.TerminationCondition?.Reset();
            this.CurrentTurn = 0;
        }
    }

    private async ValueTask<bool> TerminateIfNeededAsync(params IList<AgentMessage> incomingMessages)
    {
        if (this.TerminationCondition == null)
        {
            return false;
        }

        StopMessage? stopMessage = await this.TerminationCondition.CheckAndUpdateAsync(incomingMessages);
        if (stopMessage != null)
        {
            await this.TerminateAsync(stopMessage);

            return true;
        }

        return false;
    }

    // TODO: Figure out how to route this to the right method
    //private ValueTask ProcessNextSpeakerAsync(params IList<AgentMessage> incomingMessages)
    //    => this.ProcessNextSpeakerAsync(incomingMessages);

    private async ValueTask ProcessNextSpeakerAsync()
    {
        string nextSpeakerTopic = await this.SelectSpeakerAsync(this.MessageThread);
        await this.PublishMessageAsync(new GroupChatRequestPublish { }, nextSpeakerTopic);
    }

    public ValueTask HandleAsync(object item, MessageContext messageContext)
    {
        throw new NotImplementedException();
    }
}

File: typescript/src/agentchat/groupchat/GroupChatManagerBase.ts
import { ChatMessage, StopMessage, AgentMessage } from "../abstractions/Messages";
import { ITerminationCondition } from "../abstractions/Termination";
import { GroupChatOptions } from "./GroupChatOptions";
import { IHandle } from "../../contracts/IHandle";
import { 
    GroupChatEventBase, 
    GroupChatStart, 
    GroupChatAgentResponse, 
    GroupChatRequestPublish, 
    GroupChatMessage, 
    GroupChatTermination, 
    GroupChatReset 
} from "./Events";
import { MessageContext } from "../../contracts/MessageContext";

/**
 * Delegate type for publishing messages.
 */
type MessagePublishServicer = (event: GroupChatEventBase, topicType: string, cancellation?: AbortSignal) => Promise<void>;

/**
 * Interface for handlers that can process group chat events.
 */
export interface IGroupChatHandler {
    /**
     * Attaches a servicer for publishing messages.
     */
    attachMessagePublishServicer(servicer?: MessagePublishServicer): void;

    /**
     * Detaches the current message publishing servicer.
     */
    detachMessagePublishServicer(): void;

    /**
     * Handles group chat messages.
     */
    handleAsync(message: GroupChatStart | GroupChatAgentResponse | unknown, context: MessageContext): Promise<void>;
}

/**
 * Base class for managing group chat interactions.
 */
export abstract class GroupChatManagerBase implements IGroupChatHandler {
    protected readonly options: GroupChatOptions;
    protected messagePublishServicer?: MessagePublishServicer;
    protected messages: AgentMessage[] = []; // Changed from ChatMessage[] to AgentMessage[]
    private readonly terminationCondition?: ITerminationCondition;
    private readonly maxTurns?: number;
    private currentTurn: number = 0;

    constructor(options: GroupChatOptions) {
        this.options = options;
        this.terminationCondition = options.terminationCondition;
        this.maxTurns = options.maxTurns;
    }

    attachMessagePublishServicer(servicer?: MessagePublishServicer): void {
        this.messagePublishServicer = servicer;
    }

    detachMessagePublishServicer(): void {
        this.messagePublishServicer = undefined;
    }

    protected async publishMessageAsync(
        event: GroupChatEventBase,
        topicType: string,
        cancellation?: AbortSignal
    ): Promise<void> {
        if (this.messagePublishServicer) {
            await this.messagePublishServicer(event, topicType, cancellation);
        }
    }

    /**
     * Handles the start of a group chat session.
     */
    async handleAsync(message: GroupChatStart | GroupChatAgentResponse | unknown, context: MessageContext): Promise<void> {
        if (message instanceof GroupChatStart) {
            return this.handleStartAsync(message, context);
        } else if (message instanceof GroupChatAgentResponse) {
            return this.handleResponseAsync(message, context);
        }
        throw new Error(`Unhandled message type: ${typeof message}`);
    }

    protected abstract handleStartAsync(message: GroupChatStart, context: MessageContext): Promise<void>;
    protected abstract handleResponseAsync(message: GroupChatAgentResponse, context: MessageContext): Promise<void>;

    protected async checkTerminationConditionAsync(messages: AgentMessage[]): Promise<StopMessage | null> {
        if (this.maxTurns && ++this.currentTurn >= this.maxTurns) {
            return new StopMessage("Maximum turns reached", "system");
        }

        if (this.terminationCondition) {
            return await this.terminationCondition.checkAndUpdateAsync(messages);
        }

        return null;
    }

    protected async resetAsync(): Promise<void> {
        this.currentTurn = 0;
        this.messages = [];
        this.terminationCondition?.reset();
        
        if (this.messagePublishServicer) {
            await this.messagePublishServicer(
                new GroupChatReset(),
                this.options.groupChatTopicType
            );
        }
    }
}
